### PR TITLE
Add modular libraries to modulepath instead of classpath

### LINKED ISSUE #1045: Cannot set 'modular' per library in 'Project and External Dependencies'
Based on some examples I found on internet, I tried to add some libraries (but not all) to the modulepath, but it has no effect whatsoever.

If I run 'gradle eclipse' it creates the correct .classpath, it adds all the libraries separately, with the module-flags set as I need.  This way I can run the unit tests from eclipse.  

BUT...  if I do 'gradle refresh' from eclipse, the .classpath gets overwritten, all the libraries removed, and instead the gradle-container is added, and the information about modulepath is lost.

I need this to run the tests, I do not want to add the complete gradle-container to the modulepath, because then I would also have to add 'require junit' to the module-info, which does not make any sense for main-code.   (And eclipse cannot handle a separate module-info for the test source)

Here is what i tried:

```
boolean isModule(entry) {
  // filter java 9 modules: libs that we don't use for testing
  entry.kind == 'lib' && entry.entryAttributes.get("test") != "true"
}

eclipse {
  classpath {
    file {
      whenMerged {
        entries.findAll { isModule(it) }.each {
           it.entryAttributes['module'] = 'true'
        }
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by RuthAlk**
Fixes #1045 ?

### Context

Running junit tests from eclipse will give a 'module not found' exception if a modular library is required by the project. This commit fixes that.

### CODE REVIEW

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/GradleClasspathContainerRuntimeClasspathEntryResolver.java [hunk 1]
```diff
@@ -161,6 +163,9 @@ private void collectContainerRuntimeClasspathWithGradleScopes(IClasspathContaine
             throw new GradlePluginsRuntimeException("JavaRuntime.resolveRuntimeClasspathEntry() should not be called when Buildship is installed for Eclipse 4.8", e);
         }
     }
+    private boolean hasModuleAttribute(IClasspathEntry entry) {
+        return false && hasEnabledBooleanAttribute(IClasspathAttribute.MODULE, entry);
```
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/GradleClasspathContainerRuntimeClasspathEntryResolver.java:167]:
```suggestion
        return hasEnabledBooleanAttribute(IClasspathAttribute.MODULE, entry);
```

**Review by donat — CHANGES_REQUESTED** [review]
Looks good, I have only one suggestion.
Also, please mint the validation failure on the overview page. You'll find the description of what needs to be done here: https://wiki.eclipse.org/Development_Resources/Contributing_via_Git#via_GitHub

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/GradleClasspathContainerRuntimeClasspathEntryResolver.java (hunk 1 above)
Comment (RuthAlk) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/GradleClasspathContainerRuntimeClasspathEntryResolver.java:167]: Oh, that is silly! I had wanted to comment it out and see if it really made the difference, or whether it was already fixed. Forgot to comment it in again ... :-(

**Comment by RuthAlk** [comment]
I've modified the commit, so that it now passes the legal checks, and also made you suggested change.

**Review by donat — APPROVED** [review]
LGTM. I'm going to run this through CI and merge it if it passes.

**Review by donat — CHANGES_REQUESTED** [review]
The build failed with a compilation error. `IRuntimeClasspathEntry.MODULE_PATH` does not exist in Eclipse 4.7. Buildship supports all Eclipse versions until 4.3. Can you please take a look at that? You'll find an example of how to fix this issue with reflection in this class (i.e. `invokeJavaRuntimeResolveRuntimeClasspathEntry()`). You can test your changes by running `./gradlew build -Peclipse.version=47.

**Comment by RuthAlk** [comment]
I have added another commit so that it now uses reflection.

I tried to test it, as you suggested.  I ran "./gradlew build -Peclipse.version=43"  and after applying these changes, that compiled. I couldn't run the complete tests, though, I got a logfile with a stacktrace (see below), but I doubt that that has anything to do with my changes.  Perhaps something wrong in my environment. 

Maybe you can just check once more if the latest commit passes the CI tests?

```
!ENTRY org.eclipse.osgi 4 0 2020-12-09 19:07:25.015
!MESSAGE Startup error
!STACK 1
java.lang.RuntimeException: Exception in org.eclipse.osgi.framework.internal.core.SystemBundleActivator.start() of bundle org.eclipse.osgi.
  at org.eclipse.osgi.framework.internal.core.InternalSystemBundle.resume(InternalSystemBundle.java:233)
  at org.eclipse.osgi.framework.internal.core.Framework.launch(Framework.java:656)
  at org.eclipse.core.runtime.adaptor.EclipseStarter.startup(EclipseStarter.java:275)
  at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:177)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.base/java.lang.reflect.Method.invoke(Method.java:566)
  at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:636)
  at org.eclipse.equinox.launcher.Main.basicRun(Main.java:591)
  at org.eclipse.equinox.launcher.Main.run(Main.java:1450)
  at org.eclipse.equinox.launcher.Main.main(Main.java:1426)
Caused by: org.osgi.framework.BundleException: Exception in org.eclipse.osgi.framework.internal.core.SystemBundleActivator.start() of bundle org.eclipse.osgi.
  at org.eclipse.osgi.framework.internal.core.BundleContextImpl.startActivator(BundleContextImpl.java:734)
  at org.eclipse.osgi.framework.internal.core.BundleContextImpl.start(BundleContextImpl.java:683)
  at org.eclipse.osgi.framework.internal.core.InternalSystemBundle.resume(InternalSystemBundle.java:225)
  ... 11 more
Caused by: java.lang.IllegalStateException: The System Bundle could not be resolved: Missing Constraint: Bundle-RequiredExecutionEnvironment: J2SE-1.5
  at org.eclipse.osgi.internal.baseadaptor.BaseStorage.checkSystemState(BaseStorage.java:827)
  at org.eclipse.osgi.internal.baseadaptor.BaseStorage.getStateManager(BaseStorage.java:800)
  at org.eclipse.osgi.baseadaptor.BaseAdaptor.getState(BaseAdaptor.java:387)
  at org.eclipse.osgi.internal.baseadaptor.BaseStorage.frameworkStart(BaseStorage.java:923)
  at org.eclipse.osgi.baseadaptor.BaseAdaptor.frameworkStart(BaseAdaptor.java:250)
  at org.eclipse.osgi.framework.internal.core.SystemBundleActivator.start(SystemBundleActivator.java:60)
  at org.eclipse.osgi.framework.internal.core.BundleContextImpl$1.run(BundleContextImpl.java:711)
  at java.base/java.security.AccessController.doPrivileged(Native Method)
  at org.eclipse.osgi.framework.internal.core.BundleContextImpl.startActivator(BundleContextImpl.java:702)



```

**Comment by donat** [comment]
I'll do that.

**Review by donat — CHANGES_REQUESTED** [review]
This is sort of a big change tbh which might break the existing workspaces out there. I'm leaning towards hiding this feature in a preview flag. I've already prepared the implementation. Would you mind integrating it? https://github.com/eclipse/buildship/commit/965cdd576fe31dccb07cdb058cabbca10bb400df

**Comment by RuthAlk** [comment]
wow, that must have been a lot of work!

Yes, of course I understand about the feature-toggle.  I am not very experienced in working with eclipse-plugins, I might very well have missed some issues.

The only thing is: I cherry-picked your commit on top of mine, but I still need to read the setting somehow, to let it make a difference, but as I said, I am not very experienced with this plugin-building, how do I retrieve the value of that preference? Is there some method "getCurrentPreferences"?

Also the build failed with lots of test-failures like this: 
```
groovy.lang.GroovyRuntimeException: Could not find matching constructor for: org.eclipse.buildship.core.internal.configuration.WorkspaceConfiguration(org.eclipse.buildship.core.WrapperGradleDistribution, null, null, java.lang.Boolean, java.lang.Boolean, java.lang.Boolean, java.util.ArrayList, java.util.ArrayList, java.lang.Boolean, java.lang.Boolean)
	at org.eclipse.buildship.core.internal.configuration.BuildConfigurationTest.load build configuration overriding workspace settings(BuildConfigurationTest.groovy:245)
```

Not sure if that failing test is a problem?

**Comment by RuthAlk** [comment]
Never mind, I found the way to the preferences.

I just force-pushed my branch, rebased on master as well, I still don't know if those failing tests will be a problem, and for some reason the checks complain about legal agreements, but I do not see any difference in my author name/email. Perhaps it's your email in the second last commit?

**Comment by donat** [comment]
The validation failure is my fault. I used the wrong email address in the commits. I've updated my branch with a force-push. You can cherry-pick this commit https://github.com/eclipse/buildship/commit/50a0aef21f5696ad8b18308bb3dd69047b7df2a7.

**Review by donat — APPROVED** [review]

**Comment by donat** [comment]
@RuthAlk Thank you for the contribution!

**Comment by donat** [comment]
@RuthAlk fyi I've done some cleanup after merging your PR: https://github.com/eclipse/buildship/commit/c8128ca88a6b5750ad743bb68856cb1aa7fe04a1 https://github.com/eclipse/buildship/commit/c8128ca88a6b5750ad743bb68856cb1aa7fe04a1

**Pull request merged**